### DISCUSSION TITLE
Current conservation (Kirchoff's law) in the DFN model

### DISCUSSION THREAD

**Discussion by abhishek-appana** [pybamm-team/PyBaMM#3923] [Q&A] [answered]
Hello everyone,

My question surrounds around the conservation of current at each node and the calculation of electrode and electrolyte current densities in the DFN model. I had been poring through the source code to understand these but could not figure it out. Here is where I need help!

At each '_x_', the sum of electrode and electrolyte current densities should be equal to the input current density as per Kirchoff's law (Equation 24 in the [paper](https://doi.org/10.1149/1.2221597) authored by Doyle et al.. ) Can someone point me to where this is implemented in the DFN model? The closest I could come to this was in the `set_algebraic` function definitions in the 'full conductivity' submodel and 'full ohm' submodel. But they did not make complete sense.

Also, in these submodels, the current densities appear to be calculated based on these algebraic conditions. Is that correct? Or is there another source for these? This question arose because of the equations mentioned in the example given [here](https://docs.pybamm.org/en/stable/source/examples/notebooks/models/DFN.html). The electrolyte current density is calculated using the exchange current density in the first charge conservation equation. This felt more accurate than the algebraic conditions. However, I did not find this equation implemented anywhere in PyBaMM. Which file is it implemented in, if it is implemented? And I guess a similar equation would've been implemented for the electrode current density. What is the location of that equation?

Thank you for the help in advance!

**Comment by abhishek-appana** [comment]
Hello,

Would someone be able to help me with this?

Thanks in advance

**Reply by valentinsulzer** [reply, marked as answer]
The current equations are `div(i_e) - aj = 0`(in full_condutivity) and `div(i_s) + aj = 0` (in full_ohm). Both in `set_algebraic` of those submodels. Adding those together means `div(i_e + i_s) = 0` i.e. `i_e + i_s = constant` (in 1D) i.e. current is conserved. But we don't need to enforce `i_e + i_s = constant` explicitly, the way it is done via the algebraic equations is sufficient.

**Reply by abhishek-appana** [reply]
Hello @tinosulzer ,

Oh yes. I understand it now. Thank you so much!

Also, what about the below equation? Where is that implemented?

![image](https://github.com/pybamm-team/PyBaMM/assets/142918023/0265ad6b-58ee-4401-8d87-f3de03f38bd0)

**Reply by valentinsulzer** [reply]
That is the algebraic equation in full_conductivity. `i_e` is `concatentation(i_e_n, i_e_s, i_e_p)` and `sum_a_j` is `concatenation(a n * j_n, 0, a_p * j_p)`.

Check `basic_dfn.py` to see all the equations in one place.

**Reply by abhishek-appana** [reply]
Okay. I see them in the `basic_dfn.py`

But I cannot find these when using `dfn.py`. Can you help me find out where they are in the submodel equations? Mainly, I cannot find the definition of `sum_a_j`.
